cherab/solps/solps_plasma.py

```diff
@@ -248,55 +597,201 @@ def eirene_simulation(self):
-            raise RuntimeError("EIRENE simulation data not available for this SOLPS simulation")
+            raise RuntimeError("EIRENE simulation data not available for this SOLPS simulation.")
         else:
             return self._eirene
 
+    @eirene_simulation.setter
+    def eirene_simulation(self, value):
+        if not isinstance(value, Eirene):
+            raise ValueError('Attribute "eirene_simulation" must be an Eirene instance.')
+
+        self._eirene = value
+
+    def b2_flux_to_velocity(self, poloidal_flux, radial_flux, parallel_velocity):
+        """
+        Sets velocities of all species using B2 particle fluxes defined at cell faces.
+
+        :param ndarray poloidal_flux: Poloidal flux of atoms in s-1. Must be a 3 dimensional array of
+                                      shape (num_atoms, mesh.ny, mesh.nx).
+        :param ndarray radial_flux: Radial flux of atoms in s-1. Must be a 3 dimensional array of
+                                    shape (num_atoms, mesh.ny, mesh.nx).
+        :param ndarray parallel_velocity: Parallel velocity of atoms in m/s. Must be a 3 dimensional
+                                          array of shape (num_atoms, mesh.ny, mesh.nx).
+                                          Parallel velocity is a velocity projection on magnetic
+                                          field direction.
+        """
+        if self.b_field_cylindrical is None:
+            raise RuntimeError('Attribute "b_field_cylindrical" is not set.')
+        if self.species_density is None:
+            raise RuntimeError('Attribute "species_density" is not set.')
+
+        mesh = self.mesh
+        b = self.b_field_cylindrical
+
+        poloidal_flux = np.array(poloidal_flux, dtype=np.float64, copy=False)
+        radial_flux = np.array(radial_flux, dtype=np.float64, copy=False)
+        parallel_velocity = np.array(parallel_velocity, dtype=np.float64, copy=False)
+
+        nx = mesh.nx  # poloidal
+        ny = mesh.ny  # radial
+        ns = len(self.species_list)  # number of species
+
+        _check_shape('poloidal_flux', poloidal_flux, (ns, ny, nx))
+        _check_shape('radial_flux', radial_flux, (ns, ny, nx))
+        _check_shape('parallel_velocity', parallel_velocity, (ns, ny, nx))
+
+        poloidal_area = mesh.poloidal_area
+        radial_area = mesh.radial_area
+        leftix = mesh.neighbix[0]  # poloidal prev.
+        leftiy = mesh.neighbiy[0]
+        bottomix = mesh.neighbix[1]  # radial prev.
+        bottomiy = mesh.neighbiy[1]
+        rightix = mesh.neighbix[2]   # poloidal next.
+        rightiy = mesh.neighbiy[2]
+        topix = mesh.neighbix[3]  # radial next.
+        topiy = mesh.neighbiy[3]
+
+        # Converting s-1 --> m-2 s-1
+        poloidal_flux = np.divide(poloidal_flux, poloidal_area, out=np.zeros_like(poloidal_flux), where=poloidal_area > 0)
+        radial_flux = np.divide(radial_flux, radial_area, out=np.zeros_like(radial_flux), where=radial_area > 0)
```

Comment:
Does specifying the `out` argument here actually help? You're still having to allocate a new array, and I thought the `out` argument was only relevant if there was an existing array to put the result in to avoid a new allocation. There are a few other places below where you also do this.

Reply:
No, if I do not specify the `out` as a zero array explicitly, I'll get uninititalised values where the condition for divide is not met. From numpy.divide description: "...Note that if an uninitialized out array is created via the default out=None, locations within it where the condition is False will remain uninitialized.", [https://numpy.org/doc/stable/reference/generated/numpy.divide.html](https://numpy.org/doc/stable/reference/generated/numpy.divide.html).

Reply:
Ah, I see. Thanks for clarifying.